gui/default/syncthing/core/syncthingController.js

```diff
@@ -1679,7 +1679,11 @@ angular.module('syncthing.core')
                         $scope.currentDevice = p.data;
                         $scope.currentDevice.name = name;
                         $scope.currentDevice.deviceID = deviceID;
-                        $scope.currentDevice._editing = "add";
+                        if (deviceID) {
+                            $scope.currentDevice._editing = "pending";
```

Comment:
> When adding another value to this "kind of enum" variable, you must make sure that every other use of it will not change semantics with the new value. Did you check that?

I think so. This variable is used regarding adding new devices (both empty and pending), editing existing devices, and also editing default devices. Everything seems to be working as intended, but if you can see something wrong or a flaw in the code, please let me know.

> And I think this function should be refactored to completely avoid retrieving the discovery list at all for the case when deviceID is already provided.

Yeah, but I doubt I've got enough JS skills to do it. I did try to modify the code to move the logic further above at first, but it didn't really work as I wanted it to, so in the end I just decided to add small modifications to the existing code instead.

One problem I encountered was that if the discovery part was skipped completely in the case of pending devices, and then we added a new device from scratch (i.e. not a pending one) first, then the list would be remembered and then still present when adding a new pending device later. This means that the code still had to clear out the list on every pending folder adding. All in all, the code got messy very quickly and I wasn't really capable of fixing it cleanly.

Reply:
We're all here to learn... :wink:  You might get some inspiration from the recent changes regarding default ignores, where retrieving the setting can be delayed if the user wants to edit them in the following dialog.  In code:
https://github.com/syncthing/syncthing/blob/main/gui/default/syncthing/core/syncthingController.js#L1965

You basically have a variable with a Promise that immediately resolves initially.  If the circumstances require, you replace it with a promise that depends on the HTTP query.  The following code just calls `.then()` on the variable, which works in both cases to delay executing the given code until the promise is resolved.  Pretty elegant.  I don't see any reason why the same should not work here, your troubles were probably caused by getting confused with the Promise API and what gets executed when.

> I think so. This variable is used regarding adding new devices (both empty and pending), editing existing devices, and also editing default devices. Everything seems to be working as intended, but if you can see something wrong or a flaw in the code, please let me know.

What I meant was: Please do check every single line of the *code* where `currentDevice._editing` is accessed.  I will review that part myself as well once the logical structure is finalized.

Reply:
I've checked all the lines. Everything seems fine for me (which doesn't mean it's 100% flaw-free, but I've done as much as I'm able to). I've got to say though that I'm not going to attempt any major refactoring of the code in this commit (so that the discovery list isn't needlessly retrieved). If you or anyone else wants to work on it, please proceed (probably in a separate PR, I'd assume).

I don't think such a refactoring is exactly related to this specific PR, because the same can be said for the already existing code that builds the discovery list even when editing existing or default devices, where it's unnecessary already.

Reply:
This code is definitely terrible, mixing `success` and `then`, and returning a premise, but not the one called in the `then`. That's not on this PR though indeed, I can clean it up and skip the discovery call in case of adding a pending device in a followup unless you want to dig into it.

Reply:
I agree, it was a thing with room for improvement before this PR and it can be fixed after it.